valhalla/common/telescope_states.py

```diff
@@ -15,116 +15,188 @@
 ES_STRING_FORMATTER = "%Y-%m-%d %H:%M:%S"
 
 
-def get_es_data(query):
-    try:
-        es = Elasticsearch([settings.ELASTICSEARCH_URL])
-    except LocationValueError:
-        logger.error('Could not find host. Make sure ELASTICSEARCH_URL is set.')
-        raise ImproperlyConfigured('ELASTICSEARCH_URL')
-
-    event_data = []
-    query_size = 10000
-    data = es.search(index="telescope_events", body=query, size=query_size, scroll='1m',
-                     _source=['timestamp', 'telescope', 'enclosure', 'site', 'type', 'reason'], sort=['timestamp'])  # noqa
-    event_data.extend(data['hits']['hits'])
-    total_events = data['hits']['total']
-    events_read = min(query_size, total_events)
-    scroll_id = data.get('_scroll_id', 0)
-    while events_read < total_events:
-        data = es.scroll(scroll_id=scroll_id, scroll='1m') # noqa
-        scroll_id = data.get('_scroll_id', 0)
-        event_data.extend(data['hits']['hits'])
-        events_read += len(data['hits']['hits'])
-
-    return event_data
-
-
-def get_telescope_states(start, end, telescopes=None, sites=None, instrument_types=None):
-    telescope_to_instruments = ConfigDB().get_instrument_types_per_telescope(only_schedulable=True)
-    if not instrument_types:
-        available_telescopes = telescope_to_instruments.keys()
-    else:
-        available_telescopes = [tk for tk, insts in telescope_to_instruments.items() if
-                                any(inst in insts for inst in instrument_types)]
-
-    if not sites:
-        sites = list({telescope_key.site for telescope_key in available_telescopes})
-
-    if not telescopes:
-        telescopes = list({tk.telescope for tk in available_telescopes if tk.site in sites})
-
-    date_range_query = {
-        "query": {
-            "filtered": {
-                "filter": {
-                    "and": [
-                        {
-                            "range": {
-                                "timestamp": {
-                                    "gte": start.strftime(ES_STRING_FORMATTER),
-                                    "lte": end.strftime(ES_STRING_FORMATTER),
-                                    "format": "yyyy-MM-dd HH:mm:ss"
+class TelescopeStates(object):
+    def __init__(self, start, end, telescopes=None, sites=None, instrument_types=None):
+        self.start = start.replace(tzinfo=timezone.utc).replace(microsecond=0)
+        self.end = end.replace(tzinfo=timezone.utc).replace(microsecond=0)
+        self.instrument_types = instrument_types
+        self.available_telescopes = self._get_available_telescopes()
+
+        sites = list({telescope_key.site for telescope_key in self.available_telescopes}) \
+            if not sites else sites
+        telescopes = list({tk.telescope for tk in self.available_telescopes if tk.site in sites}) \
+            if not telescopes else telescopes
```

Comment:
The change from telescope_key to tk made the site list comprehension short enough to fit on one line, but alas, the telescopes line still goes over!